vortex-array/src/array/chunked/flatten.rs

```diff
@@ -0,0 +1,146 @@
+use arrow_buffer::{BooleanBuffer, MutableBuffer, ScalarBuffer};
+
+use vortex_dtype::{DType, match_each_native_ptype, Nullability, PType, StructDType};
+use vortex_error::{vortex_bail, VortexResult};
+
+use crate::{Array, ArrayTrait, ArrayValidity, Flattened, IntoArray};
+use crate::accessor::ArrayAccessor;
+use crate::array::bool::BoolArray;
+use crate::array::chunked::ChunkedArray;
+use crate::array::primitive::PrimitiveArray;
+use crate::array::r#struct::StructArray;
+use crate::array::varbin::builder::VarBinBuilder;
+use crate::array::varbin::VarBinArray;
+use crate::validity::{LogicalValidity, Validity};
+
+pub fn try_flatten_chunks(chunks: Vec<Array>, dtype: DType) -> VortexResult<Flattened> {
+    match &dtype {
+        // Structs can have their internal field pointers swizzled to push the chunking down
+        // one level internally without copying or decompressing any data.
+        DType::Struct(struct_dtype, _) => {
+            let struct_array = swizzle_struct_chunks(chunks.as_slice(), struct_dtype)?;
+            Ok(Flattened::Struct(struct_array))
+        }
+
+        // TODO(aduffy): can we pushdown chunks here?
+        DType::Extension(_, _) => {
+            // How can we stitch an unknown type back together?
+            todo!()
+        }
+
+        // Lists just flatten into their inner PType
+        DType::List(_, _) => {
+            todo!()
+        }
+
+        DType::Bool(nullability) => {
+            let bool_array = pack_bools(chunks.as_slice(), *nullability)?;
+            Ok(Flattened::Bool(bool_array))
+        }
+        DType::Primitive(ptype, nullability) => {
+            let prim_array = pack_primitives(chunks.as_slice(), *ptype, *nullability)?;
+            Ok(Flattened::Primitive(prim_array))
+        }
+        DType::Utf8(nullability) => {
+            let varbin_array = pack_varbin(chunks.as_slice(), dtype.clone(), *nullability)?;
+            Ok(Flattened::VarBin(varbin_array))
+        }
+        DType::Binary(nullability) => {
+            let varbin_array = pack_varbin(chunks.as_slice(), dtype.clone(), *nullability)?;
+            Ok(Flattened::VarBin(varbin_array))
+        }
+        DType::Null => {
+            vortex_bail!(ComputeError: "DType::Null cannot be flattened")
```

Comment:
Can't it? Isn't it a trivial flattening just summing up the length of each null array

Reply:
Possibly, I didn't see any NullArray encoding and there isn't a Flattened variant. Can add one as part of this, modeled after whatever Arrow is doing

Reply:
Ah yeah... I think we used to have one of these maybe? Seems like we should, in order to be zero-copy to Arrow. 

We could use ConstantEncoding with a null scalar, but the Flatten enum can't guarantee the scalar is a null value.

Reply:
https://github.com/spiraldb/vortex/blob/0e83666c8e76b7bee87838fe0d9b4a91dda7904c/vortex-array/src/arrow/array.rs#L195-L200

Reply:
This only really exists to be able to copy directly from arrow since they have null type.